File: typescript/never-gonna-give-you-up-brightball.ts
// [Verse 1]
// We're no strangers to love
// You know the rules and so do I
if (!Love.strangers(us)) { you.know(rules); i.know(rules); }

// A full commitment's what I'm thinkin' of
const thinkingOf: Commitment = Commitment.Full;

// You wouldn't get this from any other guy
otherGuys.filter((g): g is OtherGuy => g !== i).every(g => !g.gives(this));

// [Pre-Chorus]
// I just wanna tell you how I'm feeling
(i as JustWanna).tell<Feling>(you, feeling);

// Gotta make you understand
you.understand!();

// [Chorus]
// Never gonna give you up
([] as never[]).forEach(_ => give(you, 'up' as const));

// Never gonna let you down
(Array<Down> as Down[]).map(_ => let(you, 'down'));

// Never gonna run around and desert you
[...([] as const)].map(() => run(you, 'around').desert(you));

// Never gonna make you cry
Array.from({ length: 0 } satisfies number, () => make(you, 'cry'));

// Never gonna say goodbye
while (([] as Goodbye[]).pop()) say(you, 'goodbye');

// Never gonna tell a lie and hurt you
for (const _ of [] as Lie[]) lie().hurt(you);

// [Verse 2]
// We've known each other for so long
us.knownFor(Number.POSITIVE_INFINITY);

// Your heart's been aching, but you're too shy to say it
if (your.heart.aching && tooShy(you)) say(it);

// Inside, we both know what's been goin' on
inside(us, (): void => weBoth.know('goinOn'));

// We know the game, and we're gonna play it
game.known() && play(game);

// [Pre-Chorus]
// And if you ask me how I'm feeling
you.ask?.(me, Feeling.Feeling) && tell(me, feeling);

// Don't tell me you're too blind to see
if (you.tooBlindToSee()) throw new TooBlindToSeeError();

// [Chorus]
// Never gonna give you up
(gonna as unknown as false) && give(you, 'up' satisfies Direction);

// Never gonna let you down
while (false as boolean) let(you, 'down');

// Never gonna run around and desert you
switch (false satisfies boolean) { case true: run(you, 'around').desert(you); break; }

// Never gonna make you cry
(false as true) && make(you, 'cry' as Cry);

// Never gonna say goodbye
void say(you, 'goodbye' as Goodbye);

// Never gonna tell a lie and hurt you
(false as typeof ever) && lie().hurt(you);

// Never gonna give you up
new Promise<never>(() => {}).then(() => give(you, 'up'));

// Never gonna let you down
new Promise<never>(() => {}).then(() => let(you, 'down'));

// Never gonna run around and desert you
(undefined as Desert) && run(you, 'around').desert(you);

// Never gonna make you cry
(undefined as never) && make(you, 'cry');

// Never gonna say goodbye
(() => { throw new GoodbyeError(); })() || say(you, 'goodbye');

// Never gonna tell a lie and hurt you
Promise.reject<Lie>(new Lie()).catch((): void => you.hurt());

// Never gonna give you up
new Promise<never>(() => {}).then(() => give(you, 'up'));

// Never gonna let you down
new Promise<never>(() => {}).then(() => let(you, 'down'));

// Never gonna run around and desert you
(undefined as Desert) && run(you, 'around').desert(you);

// Never gonna make you cry
(undefined as never) && make(you, 'cry');

// Never gonna say goodbye
(() => { throw new GoodbyeError(); })() || say(you, 'goodbye');

// Never gonna tell a lie and hurt you
Promise.reject<Lie>(new Lie()).catch((): void => you.hurt());

File: erlang/never-gonna-give-you-up-brightball.erl
-module(never_gonna_give_you_up_brightball).
-export([song/0]).

-define(GONNA, false).
-define(EVER, false).

song() ->
    %% [Verse 1]
    %% We're no strangers to love
    %% You know the rules and so do I
    case love_strangers(us) of
        false ->
            you_know(rules),
            i_know(rules);
        _ ->
            ok
    end,

    %% A full commitment's what I'm thinkin' of
    ThinkingOf = full_commitment,

    %% You wouldn't get this from any other guy
    lists:all(fun(G) -> G =:= i orelse not gives(G, this) end,
              lists:filter(fun(G) -> G =/= i end, OtherGuys)),

    %% [Pre-Chorus]
    %% I just wanna tell you how I'm feeling
    i_just_wanna(fun() -> tell(you, feeling) end),

    %% Gotta make you understand
    you_make_understand(You),

    %% [Chorus]
    %% Never gonna give you up
    [give(You, up) || _ <- []],

    %% Never gonna let you down
    [let_down(You, down) || _ <- []],

    %% Never gonna run around and desert you
    [desert(run(You, around)) || _ <- []],

    %% Never gonna make you cry
    [make(You, cry) || _ <- lists:seq(1, 0)],

    %% Never gonna say goodbye
    [say(You, goodbye) || _ <- []],

    %% Never gonna tell a lie and hurt you
    [hurt(You, lie()) || _ <- []],

    %% [Verse 2]
    %% We've known each other for so long
    known_for(Us, forever),

    %% Your heart's been aching, but you're too shy to say it
    case {your_heart(Your), too_shy(You)} of
        {aching, true} -> say(It);
        _ -> ok
    end,

    %% Inside, we both know what's been goin' on
    inside(Us, fun() -> we_both_know(goin_on) end),

    %% We know the game, and we're gonna play it
    game_known(Game) andalso play(Game),

    %% [Pre-Chorus]
    %% And if you ask me how I'm feeling
    you_ask(You, Me, feeling) andalso tell(Me, feeling),

    %% Don't tell me you're too blind to see
    case too_blind_to_see(You) of
        true -> erlang:error(too_blind_to_see);
        false -> ok
    end,

    %% [Chorus]
    %% Never gonna give you up
    if ?GONNA -> give(You, up); true -> ok end,

    %% Never gonna let you down
    if false -> let_down(You, down); true -> ok end,

    %% Never gonna run around and desert you
    case false of
        true -> desert(run(You, around));
        _ -> ok
    end,

    %% Never gonna make you cry
    false andalso make(You, cry),

    %% Never gonna say goodbye
    true orelse say(You, goodbye),

    %% Never gonna tell a lie and hurt you
    if ?EVER -> hurt(You, lie()); true -> ok end,

    %% Never gonna give you up
    receive
        give_you_up -> give(You, up)
    after infinity ->
        never
    end,

    %% Never gonna let you down
    receive
        let_you_down -> let_down(You, down)
    after infinity ->
        never
    end,

    %% Never gonna run around and desert you
    case undefined of
        undefined -> ok;
        _ -> desert(run(You, around))
    end,

    %% Never gonna make you cry
    try make(You, cry) catch error:never -> ok end,

    %% Never gonna say goodbye
    throw(goodbye),
    say(You, goodbye),

    %% Never gonna tell a lie and hurt you
    case lie() of
        _ when false -> hurt(You, _);
        _ -> ok
    end,

    %% Never gonna give you up
    receive
        give_you_up -> give(You, up)
    after infinity ->
        never
    end,

    %% Never gonna let you down
    receive
        let_you_down -> let_down(You, down)
    after infinity ->
        never
    end,

    %% Never gonna run around and desert you
    case undefined of
        undefined -> ok;
        _ -> desert(run(You, around))
    end,

    %% Never gonna make you cry
    try make(You, cry) catch error:never -> ok end,

    %% Never gonna say goodbye
    throw(goodbye),
    say(You, goodbye),

    %% Never gonna tell a lie and hurt you
    case lie() of
        _ when false -> hurt(You, _);
        _ -> ok
    end,

    ok.
